### ISSUE TITLE
Create super resolution elevation map around known destinations

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#916]
**Is your feature request related to a problem? Please describe.**
- Elevation affects movement of a vehicle. When elevation data is of low resolution, there's quite a bit of guesswork in calculating elevation between two given points.  

**Describe the solution you desire to see**
- Create a separate elevation data double array that will encircle around known destinations such as settlements.
- The elevation data will be super high resolution, say, one data point per meter.
- They will be pseudo-randomly generated upon the start of a settlement.
- The height boundary of this double array on all four sides should be seamless with the actual MEDGR data which is also a elevation double array with global data.

**Describe the rationale**
- When returning a vehicle from mission to a well-known, well explored destination such as a settlement, settler should know exactly which route to take to get home at fastest and most efficient manner

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8180

**Additional context**
- This is somewhat related to #901, where the ability of choosing which resolution of elevation data matters.

**Comment by bevans2000** [comment]
I still believe that if we implement loading the high resolution map on the fly for a region around the target we can also achieve this purpose. 
Also we need to consolidate the various tickets around this area.

**Comment by mokun** [comment]
One problem with that high resolution map is that The map scale is only 3.705 km per pixel.

If we ever want road network or want to draw a 2.5D settlement, then we need a 1000x better elevation map.

So we pretty much need to procedurally generate one of we want to, say, having the player to do a  first person EVA walk right outside the settlement.

But for now, I could settle with, say,  just 5 m per pixel randomly generated elevation data for vehicle driving.

So, of we are to consider elevation when placing a foundation and constructing a building, no data set currently available from NASA will suffice.

**Comment by mokun** [comment]
>  implement loading the high resolution map on the fly for a region around the target 

Yes I like to implement this idea as well, in conjunction with the approach above. 

MEDGR website has even higher resolution data set but they are by regions.

We can modify TerrrainElevation class to try to read those into mars-sim. 

However, the regional area is still too big.

Even better is if we find a way to partition those big MEDGR img files into smaller regional chunks than the way they published them.

**Comment by mokun** [comment]
@bevans2000 

The current elevation caching mechanism uses *index* for reading elevation data.

What does it take to modify it to support "artificial" elevations ? 

We currently bundle `megt90n000eb.img`. The elevation data has not good enough for piloting vehicle. 

At this level 1 elevation resolution, there's a distance of 3.7 km that separate between two adjacent pixels, namely, pixel A and pixel B. 

Say, on pixel A, the elevation is 1.0 km. As a vehicle reaches pixel B, suddenly, the elevation can bump down to 0 km. Thus it can have a 1 km change of elevation.

I'd love to generate and fill in some artificial elevation data in between these two pixels.

```
//	LEVEL 0 :
//  File name: megt90n000cb.img  
//	Pixels: 1440 x 720
//	Resolution: 4 pixels per degree (or 0.25 by 0.25 degrees)
//	Scale: 14.818 km per pixel (=1/4)
//  File size: 2.025 MB
//	
//	LEVEL 1 :
//  File name: megt90n000eb.img
//	Pixels: 5760 x 2880
//	Resolution: 16 pixels per degree (or 0.0625 by 0.0625 degrees)
//	Scale: 3.705 km per pixel (=1/16)	
//  File size: 32.4 MB
//	
//	LEVEL 2 :
//  File name: megt90n000fb.img
//	Pixels: 11520 x 5760
//	Resolution: 32 pixels per degree (or 0.03125 by 0.03125 degrees)
//	Scale: 1.853 km per pixel (=1/32) 	
//  File size: 129.6 MB
```

In MEGDRCachedReader.java,
```java
    protected MEGDRCachedReader(int maximumEntries, int maxIdleMinutes) {
		
		cache = CacheBuilder.newBuilder()
								.maximumSize(maximumEntries)
								.expireAfterAccess(maxIdleMinutes, TimeUnit.MINUTES)
								.build();
	}

	/**
	 * Gets the elevation at a certain index into the map data. This will check the cached value first.
	 * 
	 * @param index Index to the data.
	 */
    protected short getElevation(int index) {
		short result = 0;
		requests++;
		Integer value = cache.getIfPresent(index);
		if (value != null) {
			result = value.shortValue();
			hits++;
		}
		else {
			result = loadElevation(index);
			cache.put(index, Integer.valueOf(result));
		}

		if (requests % 10000 == 0) {
			logger.fine("MEGDR Cache hit rate " + ((100 * hits)/requests)  + "%, requests="
						+ requests + ", size=" + cache.size());
		}
		return result;
	}
```

**Comment by bevans2000** [comment]
Interesting, I think the best way to do this is to create a new map later implementation that can convert a low res map into a high res map. So the implementation would work out the elevation as the gradient between the two points of the low res map.
Should be fairly easy to do, just some maths.

**Comment by mokun** [comment]
> So the implementation would work out the elevation as the gradient between the two points of the low res map.
Should be fairly easy to do, just some maths.

Yay. It's like getting an interpolation of the two adjacent height by just averaging them.

Here's one thing we will need to work out.

If a low resolution height map is initially available (say the one that was bundled with mars-sim binary) and the player clicked somewhere to download a higher resolution height map, then we should we do with the existing height values that may have already been populated somewhere, such as the mineral data map ?

Should we 'reset' or 'overwrite' all old height data with new ones ?

Then we'll need to write a method to first keep track of where the height data values are being populated in the first place.

This way, when a new height data is downloaded successfully, it can proceed to overwrite the old height data right away.

**Comment by mokun** [comment]
We really need to get this done.

It can solve the problem of a rover getting stuck at an outside location. 

By having a higher height resolution, it mostly avoid 
having very steep slope, by which a rover would 'fail' the terrain gradient calculation and can't find a good direction to turn. 

As a result of this, a rover will return home using, say 50%, more fuels and take much longer time.

**Comment by bevans2000** [comment]
Ok so where do we get the source info from? We cannot hold this in memory as it is too large

**Comment by mokun** [comment]
> We cannot hold this in memory as it is too large.

Can we can write the coordinate associated with its height into a JSON file ?

**Comment by mokun** [comment]
> It can solve the problem of a rover getting stuck at an outside location.

What we may first work on is to increase the resolution of the terrain right in the vicinity of a settlement.

That's where a rover can get stuck and unable to find a good direction to return home.

**Comment by bevans2000** [comment]
The format of the file needs to be binary so we can memory map it. There will be too many coordinates for a JSON representation.

**Comment by mokun** [comment]
The highest resolution of height data will narrow the terrain change to 1.9 km per pixel. See below : 

//	LEVEL 2 :
//  File name: megt90n000fb.img
//	Pixels: 11520 x 5760
//	Resolution: 32 pixels per degree (or 0.03125 by 0.03125 degrees)
//	Scale: 1.853 km per pixel (=1/32) 	
//  File size: 129.6 MB

Currently, we use 3.8 km per pixel because megt90n000eb.img is bundled with.

//	LEVEL 1 :
//  File name: megt90n000eb.img
//	Pixels: 5760 x 2880
//	Resolution: 16 pixels per degree (or 0.0625 by 0.0625 degrees)
//	Scale: 3.705 km per pixel (=1/16)	
//  File size: 32.4 MB

Maybe one way is to break up the following img file into 8x8 or 64 quadrants. 

Then we may need to develop our own tool to do so.

At least this way, it makes it easier to download just one quadrant that a settlement belongs when starting that settlement.

Then we'll need to find a way to patch the existing height data with new data over a settlement and its vicinity.

**Comment by mokun** [comment]
Alternatively, if we ever convert the whole img file into tens of thousands of json value pairs, then

maybe we can stream a few particular json pair without loading the rest of them.

Would this save the effort of having to download the entire img file ? 

Note that 129.6 MB can be compressed by xz into ~ 75 MB, I suppose. It's still not a small file to bundle with in our case.

**Comment by bevans2000** [comment]
I would not use JSON but a 2d binary file that we memory map so it can be directly referenced without having to parse.

**Comment by bevans2000** [comment]
In fact thinking about it we have to handle the spherical nature so a 2D array for the top and bottom of the globe would not work.
We still need to tile it. But what about if eachntile was a Map structure providing a lookup to the height keyed by coordinate?
But whatever we do don't commit the raw image file to GitHub. It's too big

**Comment by mokun** [comment]
> I would not use JSON but a 2d binary file that we memory map so it can be directly referenced without having to parse.

Currently, it's in a standard binary format. It has an extension of ".img".

Can we repackage it to make it more compact or truncate the files into tiles, making it more easily to stream whatever tile portions from online as needed ?
